pageserver/src/pgdatadir_mapping.rs

```diff
@@ -513,19 +537,58 @@ impl Timeline {
         version: Version<'_>,
         ctx: &RequestContext,
     ) -> Result<HashSet<RelTag>, PageReconstructError> {
-        // fetch directory listing
+        // fetch directory listing (old)
         let key = rel_dir_to_key(spcnode, dbnode);
         let buf = version.get(self, key, ctx).await?;
 
         let dir = RelDirectory::des(&buf)?;
-        let rels: HashSet<RelTag> =
+        let rels_v1: HashSet<RelTag> =
             HashSet::from_iter(dir.rels.iter().map(|(relnode, forknum)| RelTag {
                 spcnode,
                 dbnode,
                 relnode: *relnode,
                 forknum: *forknum,
             }));
 
+        if !self.get_rel_size_v2_enabled() {
+            return Ok(rels_v1);
+        }
+
+        // scan directory listing (new), merge with the old results
+        let key_range = rel_tag_sparse_key_range(spcnode, dbnode);
+        let io_concurrency = IoConcurrency::spawn_from_conf(
+            self.conf,
+            self.gate
+                .enter()
+                .map_err(|_| PageReconstructError::Cancelled)?,
+        );
+        let results = self
+            .scan(
+                KeySpace::single(key_range),
+                version.get_lsn(),
+                ctx,
+                io_concurrency,
+            )
+            .await?;
+        let mut rels = rels_v1;
+        for (key, val) in results {
+            let val = RelDirExists::decode(&val?).ok_or(PageReconstructError::Other(
+                anyhow::anyhow!("invalid reldir key"),
+            ))?;
+            if val == RelDirExists::Removed {
+                continue;
+            }
```

Comment:
Consider (debug)-asserting that it's not present in the existing results in this case.

Reply:
This is possible. I think the API design around sparse keyspace is not optimal right now, given that a key does not exist could correspond to two cases: we have a key->tombstone in the storage, or the key doesn't exist. I think this might need some refactor in the future across aux v2 and reldir v2.

Reply:
Hm? If we have a `RelDirExists::Removed` in the sparse keyspace, then there should never be an entry for it in the v1 reldir, right?

Reply:
I feel a good thing to do is to have 3 variants: RelDirExists::{Exists,Removed,NotFound}, and handle them separately to make the logic clear.

Reply:
Okay, I just realized that here the result will not contain not found, I'm messing it up with other places. I'll simply add an assertion here 🤣

Reply:
https://github.com/neondatabase/neon/pull/10593/commits/90fa52794188537b0a7a163181dfcc998bfccb26